lib/cancancan/model_adapters/neo4j_adapter.rb

```diff
@@ -26,99 +27,69 @@ def self.matches_conditions_hash?(subject, conditions)
-        matches_attribute_conditions?(atrribute_conditions, subject) &&
-          matches_associations_relations(associations_conditions, subject)
-      end
-
-      # checks if associations exists on given node
-      def self.matches_associations_relations(conditions, subject)
-        return true if conditions.blank?
-        conditions.all? do |association, value|
-          association_exists = subject.send(association).exists?
-          value ? association_exists : !association_exists
+        return false unless subject
+        conditions.all? do |key, value|
+          if (relation = base_class.associations[key])
+            match_relation_conditions(value, subject, relation)
+          else
+            property_matches?(subject, key, value)
+          end
         end
       end
 
-      def self.matches_attribute_conditions?(conditions, subject)
-        return true if conditions.blank?
+      def self.property_matches?(subject, property, value)
         if subject.is_a?(Neo4j::ActiveNode::HasN::AssociationProxy)
-          subject.where(conditions).exists?
+          subject.where(property => value).exists?
         else
-          conditions.all? do |attribute, value|
-            subject.send(attribute) == value
-          end
+          subject.send(property) == value
         end
       end
 
-      def self.associations_conditions_match?(conditions, subject, base_class)
-        return true if conditions.blank?
-        conditions.all? do |association, conditions_hash|
-          rel_length = conditions_hash.delete(:rel_length)
-          current_subject = subject.send(association, rel_length: rel_length)
-          return false unless current_subject
-          current_model = base_class.associations[association].target_class
-          all_conditions_match?(current_subject, conditions_hash, current_model)
-        end
+      def self.match_relation_conditions(conditions, subject, association)
+        rel_length = conditions.delete(:rel_length) if conditions.is_a?(Hash)
+        subject = subject.send(association.name, rel_length: rel_length)
+        return !subject.exists? if conditions.blank?
+        return subject.exists? if conditions == true
+        all_conditions_match?(subject, conditions, association.target_class)
       end
 
       private
 
       def records_for_multiple_rules
-        base_query = base_query_proxy.query
-        cypher_options = construct_cypher_options
-        cypher_options[:matches].reject! { |mt| mt.match(var_name(@model_class)) }
-        match_string = cypher_options[:matches].uniq.join(', ')
-        base_query = base_query.match(match_string) unless match_string.blank?
-        base_query
-          .proxy_as(@model_class, var_name(@model_class))
-          .where(cypher_options[:conditions])
+        query = CanCanCan::Neo4j::CypherConstructor.new(
+          construct_cypher_options
+        ).query
+        query.proxy_as(@model_class, var_name)
```

Comment:
why not nice chaining without local variable? Too many confusing `query` otherwise.